mite_selenium/__init__.py

```diff
@@ -51,32 +56,119 @@ def start(self):
-        transfer_start_timings=["requestStart"],
-        tls_time=timings["secureConnectionStart"],
-        dns_time=timings["domainLookupEnd"],
-        connect_time=timings["connectEnd"],
-        total_time=timings["duration"],
-        effective_url=timings["name"],
-    )
+    def _browser_has_timing_capabilities(self):
+        return self._context.browser.capabilities["browserName"] != "chrome"
+
+    def _is_using_tls(self, name):
+        return name.startsWith("https")
+
+    def _get_tls_timing(self, timings):
+        if self._is_using_tls(timings["name"]):
+            return timings["connectEnd"] - timings["secureConnectionStart"]
+        else:
+            logger.warning("Secure TLS connection not used, defaulting tls_time to 0")
+            return 0
+
+    def _get_tcp_timing(self, timings):
+        if self._is_using_tls(timings["name"]):
+            return timings["secureConnectionStart"] - timings["connectStart"]
+        else:
+            return timings["connectEnd"] - timings["connectStart"]
+
+    def _send_page_load_metrics(self):
+        if self._browser_has_timing_capabilities():
+            timings, *_ = self._context.browser.execute_script(
+                "return performance.getEntriesByType('navigation')"
+            )
+            protocol = timings["nextHopProtocol"]
+            if protocol != "http/1.1":
+                logger.warning(
+                    f"Timings may be inaccurate as protocol is not http/1.1: {protocol}"
+                )
+            metrics = {
+                "tcp_time": self._get_tcp_timing(timings),
+                "tls_time": self._get_tls_time_if_applicable(timings),
+                "dns_lookup_time": timings["domainLookupEnd"]
+                - timings["domainLookupStart"],
+                "time_to_first_byte": timings["responseStart"] - timings["connectionEnd"],
+                "time_to_last_byte": timings["responseEnd"] - timings["connectionEnd"],
+                "page_weight": timings["transferSize"],
+            }
+            self._context.send(
+                "http_selenium_metrics",
+                *self._extract_and_convert_metrics_to_seconds(metrics),
+            )
+
+    def _extract_and_convert_metrics_to_seconds(self, metrics):
+        converted_metrics = defaultdict()
+        for k, v in metrics:
+            converted_metrics[k] = self._convert_ms_to_seconds(v)
+        return converted_metrics
+
+    # TODO: Implement
+    def _send_javascript_metrics(self):
+        pass
+
+    def _convert_ms_to_seconds(value_ms):
+        return (value_ms / 1000) % 60
+
+    @asynccontextmanager
+    async def publish_javascript_metrics(self, transaction_name):
+        try:
+            print("Doing something for transaction")
+            yield
+        finally:
+            self._send_javascript_metrics()
+
+    def get(self, url):
+        self._context.browser.get(url)
+        self._send_page_load_metrics()
```

Comment:
```suggestion
    async def get(self, url):
        await self._context.browser.get(url)
        self._send_page_load_metrics()
```

This is how it *should* work...I hope it's true (if not we have async api design problems)

Reply:
I've changed this to be async 👍